MBS-12364, MBS-12367: Make use of AC MBIDs in website and webservice

yvanzo opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Server.
    We appreciate your time and interest in helping our project!

    Please use this template to help us review your change.

    Depending on your change, some sections may be unneeded, just remove these.
    For example, small pull requests usually don’t need the section “Action”.

    Remember that the more helpful info your pull request includes,
    the easier it is for us to understand and review your changes.

    Ensure that you’ve read through and followed the Contributing Guidelines, at
    https://github.com/metabrainz/musicbrainz-server/blob/master/CONTRIBUTING.md
-->

# Problem
<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

MBS-11456 added GIDs to artist credits in the database and MBS back-end.
Neither the website nor the webservice made these usable so far.

# Solution
<!--
    Talk about technical details, considerations, or other interesting points.
-->

## MBS-12364

For the website:
- Access the artist-credit page using MBID instead of row ID.
- Redirect row ID and old MBIDs to the canonical/current MBID.

It has some side changes as well:
- Redirect from `/mbid/<MBID>` to `/artist-credit/<MBID>` if applicable
- Support `[artist-credit:<mbid>]` link syntax in annotation
- Count ACs for MBIDs in statistics
- Move ACs cache keys to use MBIDs

## MBS-12367

For the API lookup/browse results:
- Add `id` attribute to `artist-credit` **XML** element
- Add `artist-credit-id` **JSON** key as a sibling of `artist-credit` array

# Test hints

For manually testing webpages:

- `/mbid/94eda81b-abcc-3365-9a3c-10633d06fc71`
- `/artist-credit/94eda81b-abcc-3365-9a3c-10633d06fc71`
- `/artist-credit/1599741`
- `/artist-credit/37f3d5f8-0bb7-48e2-8f82-99b73f005fcc`

  after having added an arbitrary GID redirect with:
  
  ```sql
  INSERT INTO artist_credit_gid_redirect (gid, new_id)
      VALUES ('37f3d5f8-0bb7-48e2-8f82-99b73f005fcc', 1599741);
  ```

For manually testing changes in API JSON results:

- `/ws/2/recording/52bc6f21-2c6d-48d0-be24-6ddb06a8461c?fmt=json&inc=artists`
- `/ws/2/recording/52bc6f21-2c6d-48d0-be24-6ddb06a8461c?fmt=json&inc=artist-credits`
- `/ws/2/release/a7034f6f-d038-405a-8692-59f7530f6958?fmt=json&inc=artists`
- `/ws/2/release/a7034f6f-d038-405a-8692-59f7530f6958?fmt=json&inc=artist-credits`
- `/ws/2/release-group/7c4cab8d-dead-3870-b501-93c90fd0a580?fmt=json&inc=artists`
- `/ws/2/release-group/7c4cab8d-dead-3870-b501-93c90fd0a580?fmt=json&inc=artist-credits`

For manually testing changes in API XML results:

- `/ws/2/recording/52bc6f21-2c6d-48d0-be24-6ddb06a8461c?fmt=xml&inc=artists`
- `/ws/2/recording/52bc6f21-2c6d-48d0-be24-6ddb06a8461c?fmt=xml&inc=artist-credits`
- `/ws/2/release/a7034f6f-d038-405a-8692-59f7530f6958?fmt=xml&inc=artists`
- `/ws/2/release/a7034f6f-d038-405a-8692-59f7530f6958?fmt=xml&inc=artist-credits`
- `/ws/2/release-group/7c4cab8d-dead-3870-b501-93c90fd0a580?fmt=xml&inc=artists`
- `/ws/2/release-group/7c4cab8d-dead-3870-b501-93c90fd0a580?fmt=xml&inc=artist-credits`

I didn’t manually test any browse request, however they use the same code as lookup requests to serialize artist credits, and CI tests have been updated for both lookup and browse requests.

## Propsective questions to reviewers

1. Should artist credits be added to the sitemap?
2. Should we start logging artist credits on deletion?
3. Should artist credit MBIDs (and redirects) be dumped too?


# Todolist for author
<!--
    The tasks you have to do to get your change ready for review. Use this if
    you draft a pull request. Mark done tasks with an [x] as you progress. See
    https://github.blog/2019-02-14-introducing-draft-pull-requests/
-->

* [ ] Have a release plan for SEARCH-690 (which is half implemented already)
* [ ] Update `MB_SOLR_TAG` to have an `mmd-schema` version that validates XML output in CI tests
* [ ] Display “Artist credits” count in statistics (since AC MBIDs are now counted)
* [ ] Add “Details” tab (which likely conflicts with refactoring for MBS-12552)
* [ ] Update the [MusicBrainz API (lookup/browse) examples](https://wiki.musicbrainz.org/MusicBrainz_API/Examples) in XML and JSON
* [ ] Update the [MusicBrainz API Search](https://wiki.musicbrainz.org/MusicBrainz_API/Search) examples in XML and JSON


# Actions after merge
<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

1. Release and deploy changes for SEARCH-690 prior to releasing MBS
2. Make updated WikiDocs pages visible on MusicBrainz website

reosarevok commented:
> Should artist credits be added to the sitemap?

Are you adding any JSON-LD to them? If not, probably not? But let's see what @mwiencek thinks.

> Should we start logging artist credits on deletion?

I honestly didn't even remember we have `deleted_entity` since we barely ever use it - but I don't see why we wouldn't do something like this for ACs too. Looking at a sample row, I imagine the same thing could be done with ACs - unsure what we'd store, but at least the gid, type and full text name, if not artist gids too?:
` 1e4690df-5693-48c2-a80c-eb1551d7a867 | {"entity_gid": "1e4690df-5693-48c2-a80c-eb1551d7a867", "entity_type": "artist", "last_known_name": "Ernst Rolff", "last_known_comment": ""}                                               | 2013-07-22 13:00:18.742424+00
`

> Should artist credit MBIDs (and redirects) be dumped too?

Aren't they already dumped? The `artist_credit_gid_redirect` table is already in `@CORE_TABLE_LIST`.

reosarevok commented:
For `_get_hash_by_keys`, would it make sense to just not return a hash? `find_by_ids` does `values %{ $self->get_by_ids(@$ids) }` which is the same reasoning @mwiencek had in 895e36e09905b2ca1d6fcf592529c6543cf0f3c9 when he made `_get_by_keys` for `Data::Entity` return a list.

reosarevok commented on lib/MusicBrainz/Server/Data/ArtistCredit.pm:
```diff
@@ -12,11 +12,17 @@ use MusicBrainz::Server::Data::Utils qw( object_to_ids placeholders load_subobje
+with 'MusicBrainz::Server::Data::Role::EntityModelClass';
+with 'MusicBrainz::Server::Data::Role::GetByGID';
+with 'MusicBrainz::Server::Data::Role::MainTable';
+# Follows GetByGID and MainTable as it requires 'get_by_gid' and '_main_table';
```
Have you tried just consuming them all in one `with` call instead? I did that for all of `Data::` in https://github.com/metabrainz/musicbrainz-server/pull/2751 and it didn't seem to cause issues (while avoiding the need to order things in weird and specific ways) :)

yvanzo replied:
No I just followed the current coding style for now even though it isn’t perfect. Changing it isn’t straightforward either. For example “the need of order things” is still there in the case when two roles like `GIDEntityCache` and `GIDRedirect` goes around the same method `get_by_gid`.

reosarevok commented on lib/MusicBrainz/Server/WebService/Serializer/JSON/2/Utils.pm:
```diff
@@ -196,6 +197,30 @@ sub serialize_annotation {
+        }
+    } @{ $artist_credit->{names} }];
+
+    $into->{'artist-credit-id'} = $artist_credit->{gid};
```
This seems good for now - should we have a ticket to change AC to an object whenever we do breaking changes for ws/2.5 or ws/3 or whatever we end up calling it?

yvanzo replied:
It is covered by MBS-9829 already.

reosarevok reviewed (commented):
There's also a typo "seralization" in the commit message for 26aba5bc5fe46d8ee2412a3b2a2c1af19cec1b56

Thanks for working on this - it looks sensible to me so far (tested manually based on your suggestions too).

yvanzo commented:
Thanks for your comments I moved the third prospective question about sample data dump to my todolist, and fixed the typo in the last commit message.

> Are you adding any JSON-LD to them?

I didn’t think about it.

About JSON-LD, I checked that our currently produced JSON-LD doesn’t include any artist credit data and thus doesn’t need to be updated with AC MBIDs.

> For `_get_hash_by_keys`, would it make sense to just not return a hash?

There are wo reasons for returning a hash:
- The hash is directly used and returned by `get_by_ids` .
- The current implementation builds a hash anyway to set artist count for each artist credit (which isn’t needed in the default `get_by_keys`).
